pkg/webhook/statefulset/validating/statefulset_validation.go

```diff
@@ -358,3 +363,90 @@ func validateTemplateInPlaceOnly(oldTemp, newTemp *v1.PodTemplateSpec) error {
 
 	return nil
 }
+
+// ValidateVolumeClaimTemplateUpdate tests if only size expand when sc allow expansion.
+func ValidateVolumeClaimTemplateUpdate(c client.Client, sts, oldSts *appsv1beta1.StatefulSet) field.ErrorList {
+	if sts.Spec.VolumeClaimUpdateStrategy.Type == "" ||
+		sts.Spec.VolumeClaimUpdateStrategy.Type == appsv1beta1.OnPVCDeleteVolumeClaimUpdateStrategyType {
+		return nil
+	}
+	if len(sts.Spec.VolumeClaimTemplates) != len(oldSts.Spec.VolumeClaimTemplates) {
+		return field.ErrorList{field.Invalid(field.NewPath("spec", "volumeClaimTemplates"), sts.Spec.VolumeClaimTemplates, "volumeClaimTemplate can not be added or deleted when OnRollingUpdate")}
+	}
+
+	name2Template := make(map[string]*v1.PersistentVolumeClaim)
+	for i := range oldSts.Spec.VolumeClaimTemplates {
+		name2Template[oldSts.Spec.VolumeClaimTemplates[i].Name] = &oldSts.Spec.VolumeClaimTemplates[i]
+	}
+
+	var err error
+	for _, template := range sts.Spec.VolumeClaimTemplates {
+		templateIdStr := fmt.Sprintf("volumeClaimTemplates[%v]", template.Name)
+		oldTemplate, exist := name2Template[template.Name]
+		if !exist {
+			return field.ErrorList{field.Forbidden(field.NewPath("spec", templateIdStr, "name"), "volumeClaimTemplate name can not be modified")}
+		}
+
+		matched, resizeOnly := pvc.CompareWithCheckFn(oldTemplate, &template, isPVCResizeOnly)
+		if matched {
+			continue
+		}
+		if !resizeOnly {
+			return field.ErrorList{field.Invalid(field.NewPath("spec", templateIdStr), template, "volumeClaimTemplate can not be modified when OnRollingUpdate")}
+		}
+		// check if sc allow volume expand
+		var sc *storagev1.StorageClass
+		scName := template.Spec.StorageClassName
+		if scName == nil {
+			// nil scName means using default storage class
+			sc, err = GetDefaultStorageClass(c)
+			if err != nil {
+				return field.ErrorList{field.Invalid(field.NewPath("spec", templateIdStr, "spec", "storageClassName"), "nil", "can not list storage class")}
+			}
+			//	if there is no default sc, skip check
+			if sc == nil {
+				continue
+			}
+		} else {
+			sc = &storagev1.StorageClass{}
+			err = c.Get(context.TODO(), client.ObjectKey{Name: *scName}, sc)
+			if err != nil || sc == nil {
+				return field.ErrorList{field.Invalid(field.NewPath("spec", templateIdStr, "spec", "storageClassName"), *scName, "can not get sc")}
+			}
+		}
+		if sc.AllowVolumeExpansion != nil && !*sc.AllowVolumeExpansion {
+			return field.ErrorList{field.Forbidden(field.NewPath("spec", templateIdStr, "spec", "resources", "requests", "storage"),
+				fmt.Sprintf("sc %v disallow volume expansion", sc.Name))}
```

Comment:
"sc %v disallow volume expansion" -> "storage class %v does not support volume expansion"